### PR TITLE
Generalize 2-arg `eigen` towards `AbstractMatrix`

### LINKED ISSUE #953: LinearAlgebra eigen errors when given Diagonal
When I pass matrices of type `Diagonal` to `eigen`, I get errors. This should work, right?
<details>

<summary>Demonstrating the issue</summary>

```
julia> using LinearAlgebra

julia> A = Diagonal([1, 2])
2×2 Diagonal{Int64, Vector{Int64}}:
 1  ⋅
 ⋅  2

julia> eigen(A, A)
ERROR: MethodError: no method matching eigen!(::Diagonal{Float64, Vector{Float64}}, ::Diagonal{Float64, Vector{Float64}})
Closest candidates are:
  eigen!(::Union{Hermitian{T, S}, Hermitian{Complex{T}, S}, Symmetric{T, S}}, ::AbstractMatrix{T}; sortby) where {T<:Number, S<:(StridedMatrix{T} where T)} at C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\symmetriceigen.jl:170
Stacktrace:
 [1] eigen(A::Diagonal{Int64, Vector{Int64}}, B::Diagonal{Int64, Vector{Int64}}; kws::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ LinearAlgebra C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\eigen.jl:511
 [2] eigen(A::Diagonal{Int64, Vector{Int64}}, B::Diagonal{Int64, Vector{Int64}})
   @ LinearAlgebra C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\eigen.jl:509
 [3] top-level scope
   @ REPL[3]:1

julia> eigen(A|>Matrix, A)
ERROR: MethodError: no method matching eigen!(::Matrix{Float64}, ::Diagonal{Float64, Vector{Float64}})
Closest candidates are:
  eigen!(::StridedMatrix{T}; permute, scale, sortby) where T<:Union{Float32, Float64} at C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\eigen.jl:149
  eigen!(::StridedMatrix{T}, ::StridedMatrix{T}; sortby) where T<:Union{Float32, Float64} at C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\eigen.jl:436
  eigen!(::Union{Hermitian{T, S}, Hermitian{Complex{T}, S}, Symmetric{T, S}}, ::AbstractMatrix{T}; sortby) where {T<:Number, S<:(StridedMatrix{T} where T)} at C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\symmetriceigen.jl:170
Stacktrace:
 [1] eigen(A::Matrix{Int64}, B::Diagonal{Int64, Vector{Int64}}; kws::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ LinearAlgebra C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\eigen.jl:511
 [2] eigen(A::Matrix{Int64}, B::Diagonal{Int64, Vector{Int64}})
   @ LinearAlgebra C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\eigen.jl:509
 [3] top-level scope
   @ REPL[4]:1

julia> eigen(A, A|>Matrix)
ERROR: MethodError: no method matching eigen!(::Diagonal{Float64, Vector{Float64}}, ::Matrix{Float64})
Closest candidates are:
  eigen!(::StridedMatrix{T}, ::StridedMatrix{T}; sortby) where T<:Union{Float32, Float64} at C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\eigen.jl:436
  eigen!(::Union{Hermitian{T, S}, Hermitian{Complex{T}, S}, Symmetric{T, S}}, ::AbstractMatrix{T}; sortby) where {T<:Number, S<:(StridedMatrix{T} where T)} at C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\symmetriceigen.jl:170
Stacktrace:
 [1] eigen(A::Diagonal{Int64, Vector{Int64}}, B::Matrix{Int64}; kws::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ LinearAlgebra C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\eigen.jl:511
 [2] eigen(A::Diagonal{Int64, Vector{Int64}}, B::Matrix{Int64})
   @ LinearAlgebra C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\eigen.jl:509
 [3] top-level scope
   @ REPL[5]:1

julia> eigen(A|>Matrix, A|>Matrix)
GeneralizedEigen{Float64, Float64, Matrix{Float64}, Vector{Float64}}
values:
2-element Vector{Float64}:
 0.9999999999999998
 1.0
vectors:
2×2 Matrix{Float64}:
 0.0       1.0
 0.707107  0.0

julia> versioninfo()
Julia Version 1.8.0
Commit 5544a0fab7 (2022-08-17 13:38 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: 8 × Intel(R) Core(TM) i5-8265U CPU @ 1.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, skylake)
  Threads: 8 on 8 virtual cores
```
</details>

### LINKED ISSUE #46776: LinearAlgebra eigen errors when given Diagonal
When I pass matrices of type `Diagonal` to `eigen`, I get errors. This should work, right?
<details>

<summary>Demonstrating the issue</summary>

```
julia> using LinearAlgebra

julia> A = Diagonal([1, 2])
2×2 Diagonal{Int64, Vector{Int64}}:
 1  ⋅
 ⋅  2

julia> eigen(A, A)
ERROR: MethodError: no method matching eigen!(::Diagonal{Float64, Vector{Float64}}, ::Diagonal{Float64, Vector{Float64}})
Closest candidates are:
  eigen!(::Union{Hermitian{T, S}, Hermitian{Complex{T}, S}, Symmetric{T, S}}, ::AbstractMatrix{T}; sortby) where {T<:Number, S<:(StridedMatrix{T} where T)} at C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\symmetriceigen.jl:170
Stacktrace:
 [1] eigen(A::Diagonal{Int64, Vector{Int64}}, B::Diagonal{Int64, Vector{Int64}}; kws::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ LinearAlgebra C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\eigen.jl:511
 [2] eigen(A::Diagonal{Int64, Vector{Int64}}, B::Diagonal{Int64, Vector{Int64}})
   @ LinearAlgebra C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\eigen.jl:509
 [3] top-level scope
   @ REPL[3]:1

julia> eigen(A|>Matrix, A)
ERROR: MethodError: no method matching eigen!(::Matrix{Float64}, ::Diagonal{Float64, Vector{Float64}})
Closest candidates are:
  eigen!(::StridedMatrix{T}; permute, scale, sortby) where T<:Union{Float32, Float64} at C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\eigen.jl:149
  eigen!(::StridedMatrix{T}, ::StridedMatrix{T}; sortby) where T<:Union{Float32, Float64} at C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\eigen.jl:436
  eigen!(::Union{Hermitian{T, S}, Hermitian{Complex{T}, S}, Symmetric{T, S}}, ::AbstractMatrix{T}; sortby) where {T<:Number, S<:(StridedMatrix{T} where T)} at C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\symmetriceigen.jl:170
Stacktrace:
 [1] eigen(A::Matrix{Int64}, B::Diagonal{Int64, Vector{Int64}}; kws::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ LinearAlgebra C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\eigen.jl:511
 [2] eigen(A::Matrix{Int64}, B::Diagonal{Int64, Vector{Int64}})
   @ LinearAlgebra C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\eigen.jl:509
 [3] top-level scope
   @ REPL[4]:1

julia> eigen(A, A|>Matrix)
ERROR: MethodError: no method matching eigen!(::Diagonal{Float64, Vector{Float64}}, ::Matrix{Float64})
Closest candidates are:
  eigen!(::StridedMatrix{T}, ::StridedMatrix{T}; sortby) where T<:Union{Float32, Float64} at C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\eigen.jl:436
  eigen!(::Union{Hermitian{T, S}, Hermitian{Complex{T}, S}, Symmetric{T, S}}, ::AbstractMatrix{T}; sortby) where {T<:Number, S<:(StridedMatrix{T} where T)} at C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\symmetriceigen.jl:170
Stacktrace:
 [1] eigen(A::Diagonal{Int64, Vector{Int64}}, B::Matrix{Int64}; kws::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ LinearAlgebra C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\eigen.jl:511
 [2] eigen(A::Diagonal{Int64, Vector{Int64}}, B::Matrix{Int64})
   @ LinearAlgebra C:\Users\Dennis Bal\.julia\juliaup\julia-1.8.0+0.x64\share\julia\stdlib\v1.8\LinearAlgebra\src\eigen.jl:509
 [3] top-level scope
   @ REPL[5]:1

julia> eigen(A|>Matrix, A|>Matrix)
GeneralizedEigen{Float64, Float64, Matrix{Float64}, Vector{Float64}}
values:
2-element Vector{Float64}:
 0.9999999999999998
 1.0
vectors:
2×2 Matrix{Float64}:
 0.0       1.0
 0.707107  0.0

julia> versioninfo()
Julia Version 1.8.0
Commit 5544a0fab7 (2022-08-17 13:38 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: 8 × Intel(R) Core(TM) i5-8265U CPU @ 1.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, skylake)
  Threads: 8 on 8 virtual cores
```
</details>

### PR DESCRIPTION
**Pull request by dkarrasch**
Fixes JuliaLang/LinearAlgebra.jl#953, and cleans up `symmetriceigen.jl` code a bit (without functional changes).

### CODE REVIEW

**Comment by pcjentsch** [comment]
As I mentioned in my reply in #46776, with this implementation,

`eigen(A::Diagonal,B::Diagonal) != eigen(Matrix(A),B::Diagonal) != eigen(Matrix(A),Matrix(B))`

despite these all representing the same matrices, but I think this would hold for the other matrix types. They're all correct of course. Does that matter?

**Comment by dkarrasch** [comment]
As long as you get a correct answer, I'm not sure a user would bother whether they would get the same result if they chose a different (i.e. dense) representation. Provided correctness, the next goal is performance, so simply densifying doesn't sound like an appealing option to me.

**Comment by jishnub** [comment]
While this is being addressed, I wonder if specializing cases like `eigen(A::AbstractMatrix, B::Symmetric)` are worth it, where `B` is positive-definite? Such cases arise at times (I am working with one such case in a fluid dynamics application), and this may be transformed to a standard eigenvalue problem by performing a Cholesky decomposition of `B`, which would lead to a significant speedup over evaluating `eigen(A, Matrix(S))`.

An example for 1000x1000 matrices:
```julia
julia> A = rand(1000,1000);

julia> X = rand(1000,1000);

julia> S2 = Symmetric(X'*X);

julia> @time λ1, v1 = eigen(A, Matrix(S2));
  4.999689 seconds (25 allocations: 45.953 MiB)

julia> @time λ2, v2 = begin
           C = cholesky(S2)
           eigen(C.L\(A*inv(C.U)))
       end;;
  0.399636 seconds (37 allocations: 71.680 MiB)
```
Currently, this method errors,
```julia
julia> eigen(A, S2);
ERROR: MethodError: no method matching eigen!(::Matrix{Float64}, ::Symmetric{Float64, Matrix{Float64}})
```
so limiting it to the special case of positive definite matrices might not be too bad, I guess?

**Comment by dkarrasch** [comment]
We have

https://github.com/JuliaLang/julia/blob/276af848511510861f8d0e6c1e69e15e4c646e3e/stdlib/LinearAlgebra/src/symmetriceigen.jl#L170-L175

which does exactly what you propose, but the method signature don't lead there for your example. I'll try to come up with some "promotion mechanism" that would help your case.

**Comment by dkarrasch** [comment]
Actually, can you help me benchmark the `UtiAUi` thing and compare to `rdiv!(ldiv!(U', parent(A)), U)`? I must have made a mistake, but on my machine `UtiAUi!` is not allocation-free, and performance is just horrible.

File: stdlib/LinearAlgebra/src/diagonal.jl [hunk 1]
```diff
@@ -762,6 +762,29 @@ function eigen(D::Diagonal; permute::Bool=true, scale::Bool=true, sortby::Union{
     end
     Eigen(λ, evecs)
 end
+function eigen(Da::Diagonal, Db::Diagonal; sortby::Union{Function,Nothing}=nothing)
+    if any(!isfinite, Da.diag) || any(!isfinite, Db.diag)
+        throw(ArgumentError("matrices contain Infs or NaNs"))
+    end
+    if !isposdef(Db)
+        throw(ArgumentError("right-hand side diagonal matrix is not positive definit"))
```
Comment (antoine-levitt) [inline stdlib/LinearAlgebra/src/diagonal.jl:770]: definite (several occurences)

Comment (antoine-levitt) [reply stdlib/LinearAlgebra/src/diagonal.jl:770]: don't errors have dots at the end?

**Review by antoine-levitt — COMMENTED** [review]
I'm incompetent to review the dispatch logic (which feels a bit convoluted, but then most of LinearAlgebra is...), but otherwise looks good to me. I've been wanting eigen for more general types for a while so this is great

**Comment by antoine-levitt** [comment]
In particular, it looks like it does eigen of tridiagonal matrices, which is awesome (this comes up all the time for quick & dirty 1D PDE examples)

File: stdlib/LinearAlgebra/src/diagonal.jl (hunk 1 above)
Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/diagonal.jl:770]: Most errors don't have a dot at the end, and if is has, we should probably start with a capital letter. You're right about the invertibility. That's what I had first, actually, so I'll revert to `any(iszero, D.diag)`.

**Comment by dkarrasch** [comment]
> In particular, it looks like it does eigen of tridiagonal matrices, which is awesome (this comes up all the time for quick & dirty 1D PDE examples)

Well, but it's not specialized. It densifies the `Tridiagonal` lhs, but keeps diagonal rhs.

**Comment by antoine-levitt** [comment]
Yeah I don't mind the fact it's not specialized, it's nice that it works at all!

**Review by andreasnoack — APPROVED** [review]

**Pull request merged**